crates/revm/src/evm_impl.rs

```diff
@@ -282,7 +282,7 @@ impl<'a, GSPEC: Spec, DB: Database, const INSPECT: bool> EVMImpl<'a, GSPEC, DB,
             acc_caller.info.balance = acc_caller
                 .info
                 .balance
-                .saturating_add(effective_gas_price * U256::from(gas.remaining() + gas_refunded));
+                .saturating_add(self.data.env.tx.gas_price * U256::from(gas.remaining() + gas_refunded));
```

Comment:
> @rakita I've implemented the change but looks like the tests are still failing. Maybe I misunderstood what you meant?

I said the incorrect thing, didn't take into account this that, in finalization, we need to return the difference of `(gas_price-effective_price)*gas_limit`

```suggestion
                .saturating_add((self.data.env.tx.gas_price-effective_gas_price) * self.data.env.tx.gas_limit)
                .saturating_add(effective_gas_price * U256::from(gas.remaining() + gas_refunded));
```

Reply:
I'm still not quite sure what you mean. I tried applying this change and the tests are still failing. 
Also when checking Geth it looks like gasPrice is used? https://github.com/ethereum/go-ethereum/blob/3a79a99f804c4bd2002fe52768c965c4be8106ea/core/state_transition.go#L400

Reply:
At the start:
we have done it as: `balance -= min(base_fee+tip,max_gas_price)*gas_limit` => `balance -= effective*gas_limit`
and now we are doing: `balance -= max_gas_price * gas_limit`

In finalize we need to return it back:
previously it was: `balance += effective*(remaining_gas+gas_refund)`
but now it needs to take into account additional gas that we subbed, and it would become: 
`balance +=(gas_price-effective_price)*gas_limit+effective*(remaining-refunded)`


it can be done as you did below, as do this only for the balance check. Not sure why it panics, will need to look at at detail

Reply:
aha gotcha. That makes sense, thanks :D

Geth does it the way I did it below by only subtracting the effective price from the account: https://github.com/ethereum/go-ethereum/blob/0c9eb8c9a47285526078beef4ab5aed5a90ab938/core/state_transition.go#L217 so this seems like the best way to go imo? *note `st.gasPrice` here is effective_gas_price: https://github.com/ethereum/go-ethereum/blob/0c9eb8c9a47285526078beef4ab5aed5a90ab938/core/types/transaction.go#L633

Reply:
yeah, I agree it is better to do as your latest, it is best to be aligned with geth with this kind of sensitive things.